libp2p-identify/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+	/// Protocols supported by the remote.
+	pub protocols: Vec<String>,
+}
+
+impl<C> ConnectionUpgrade<C> for IdentifyProtocol
+    where C: AsyncRead + AsyncWrite + 'static
+{
+	type NamesIter = iter::Once<(Bytes, Self::UpgradeIdentifier)>;
+	type UpgradeIdentifier = ();
+
+	#[inline]
+	fn protocol_names(&self) -> Self::NamesIter {
+		iter::once((Bytes::from("/ipfs/id/1.0.0"), ()))
+	}
+
+	type Output = Option<IdentifyInfo>;
+	type Future = Box<Future<Item = Self::Output, Error = IoError>>;
+
+	fn upgrade(self, socket: C, _: (), ty: Endpoint) -> Self::Future {
+		// TODO: use jack's varint library instead
+		let socket = length_delimited::Builder::new().length_field_length(1).new_framed(socket);
+
+		match ty {
+			Endpoint::Dialer => {
+				let future = socket.into_future()
+				                   .map(|(msg, _)| msg)
+				                   .map_err(|(err, _)| err)
+				                   .and_then(|msg| if let Some(msg) = msg {
+					Ok(Some(parse_proto_msg(msg)?))
+				} else {
+					Ok(None)
+				});
+
+				Box::new(future) as Box<_>
+			}
+
+			Endpoint::Listener => {
+				let info = self.information;
+
+				let listen_addrs = info.listen_addrs
+				                       .into_iter()
+				                       .map(|addr| addr.to_string().into_bytes())
+				                       .collect();
+
+				let mut message = structs_proto::Identify::new();
+				message.set_agentVersion(info.agent_version);
+				message.set_protocolVersion(info.protocol_version);
+				message.set_publicKey(info.public_key);
+				message.set_listenAddrs(listen_addrs);
+				message.set_observedAddr(info.observed_addr.to_string().into_bytes());
+				message.set_protocols(RepeatedField::from_vec(info.protocols));
+
+				let bytes = message.write_to_bytes().expect("we control the protobuf message");
+				let future = socket.send(bytes).map(|_| None);
+				Box::new(future) as Box<_>
+			}
+		}
+	}
+}
+
+// Turns a protobuf message into an `IdentifyInfo`. If something bad happens, turn it into
+// an `IoError`.
+fn parse_proto_msg(msg: BytesMut) -> Result<IdentifyInfo, IoError> {
+	match protobuf_parse_from_bytes::<structs_proto::Identify>(&msg) {
+		Ok(mut msg) => {
+			let listen_addrs = {
+				let mut addrs = Vec::new();
+				for addr in msg.take_listenAddrs().into_iter() {
+					addrs.push(bytes_to_multiaddr(addr)?);
+				}
+				addrs
+			};
+
+			let observed_addr = bytes_to_multiaddr(msg.take_observedAddr())?;
+
+			Ok(IdentifyInfo {
+				public_key: msg.take_publicKey(),
+				protocol_version: msg.take_protocolVersion(),
+				agent_version: msg.take_agentVersion(),
+				listen_addrs: listen_addrs,
+				observed_addr: observed_addr,
+				protocols: msg.take_protocols().into_vec(),
+			})
+		}
+
+		Err(err) => {
+			Err(IoError::new(IoErrorKind::InvalidData, err))
+		}
+	}
+}
+
+// Turn a `Vec<u8>` into a `Multiaddr`. If something bad happens, turn it into an `IoError`.
+fn bytes_to_multiaddr(bytes: Vec<u8>) -> Result<Multiaddr, IoError> {
```

Comment:
This would be nicer if it was expressed as a single expression using `map_err` and `and_then`

Reply:
Cannot do because I use `?` at some point.

Reply:
No you don't